kvdb/postgres/readwrite_cursor.go

```diff
@@ -0,0 +1,184 @@
+	copy(c.currKey, key)
+
+	return key, value
+}
+
+// Last positions the cursor at the last key/value pair and returns the
+// pair.
+func (c *readWriteCursor) Last() ([]byte, []byte) {
+	var (
+		key   []byte
+		value []byte
+	)
+	err := c.bucket.tx.tx.QueryRowContext(
+		context.TODO(),
+		"SELECT key, value FROM "+c.bucket.table+" WHERE "+parentSelector(c.bucket.id)+" ORDER BY key DESC LIMIT 1",
+	).Scan(&key, &value)
+
+	if err == sql.ErrNoRows {
+		return nil, nil
+	}
+
+	// Copy current key to prevent modification by the caller.
+	c.currKey = make([]byte, len(key))
+	copy(c.currKey, key)
+
+	return key, value
+}
+
+// Next moves the cursor one key/value pair forward and returns the new
+// pair.
+//
+// TODO: Optimize this by reading ahead in pages.
+func (c *readWriteCursor) Next() ([]byte, []byte) {
+	var (
+		key   []byte
+		value []byte
+	)
+	err := c.bucket.tx.tx.QueryRowContext(
+		context.TODO(),
+		"SELECT key, value FROM "+c.bucket.table+" WHERE "+parentSelector(c.bucket.id)+" AND key>$1 ORDER BY key LIMIT 1",
+		c.currKey,
+	).Scan(&key, &value)
+
+	if err == sql.ErrNoRows {
+		return nil, nil
+	}
+
+	// Copy current key to prevent modification by the caller.
+	c.currKey = make([]byte, len(key))
+	copy(c.currKey, key)
+
+	return key, value
+}
+
+// Prev moves the cursor one key/value pair backward and returns the new
+// pair.
+func (c *readWriteCursor) Prev() ([]byte, []byte) {
+	var (
+		key   []byte
+		value []byte
+	)
+	err := c.bucket.tx.tx.QueryRowContext(
+		context.TODO(),
+		"SELECT key, value FROM "+c.bucket.table+" WHERE "+parentSelector(c.bucket.id)+" AND key<$1 ORDER BY key DESC LIMIT 1",
+		c.currKey,
+	).Scan(&key, &value)
+
+	if err == sql.ErrNoRows {
+		return nil, nil
+	}
+
+	// Copy current key to prevent modification by the caller.
+	c.currKey = make([]byte, len(key))
+	copy(c.currKey, key)
+
+	return key, value
+}
+
+// Seek positions the cursor at the passed seek key.  If the key does
+// not exist, the cursor is moved to the next key after seek.  Returns
+// the new pair.
+func (c *readWriteCursor) Seek(seek []byte) ([]byte, []byte) {
+	// Return nil if trying to seek to an empty key.
+	if seek == nil {
+		return nil, nil
+	}
+
+	var (
+		key   []byte
+		value []byte
+	)
+	err := c.bucket.tx.tx.QueryRowContext(
+		context.TODO(),
+		"SELECT key, value FROM "+c.bucket.table+" WHERE "+parentSelector(c.bucket.id)+" AND key>=$1 ORDER BY key LIMIT 1",
+		seek,
+	).Scan(&key, &value)
+
+	if err == sql.ErrNoRows {
+		return nil, nil
+	}
+
+	// Copy current key to prevent modification by the caller.
+	c.currKey = make([]byte, len(key))
+	copy(c.currKey, key)
+
+	return key, value
+}
+
+// Delete removes the current key/value pair the cursor is at without
+// invalidating the cursor.  Returns ErrIncompatibleValue if attempted
+// when the cursor points to a nested bucket.
+func (c *readWriteCursor) Delete() error {
+	deleteKey := c.currKey
+
+	var key []byte
+	err := c.bucket.tx.tx.QueryRowContext(
```

Comment:
Now it should be in line with bolt, but needs double-checking

Reply:
So I still think this is wrong and should be as pointed in the change in that commit.
A simple pseudo code example will show what I mean:
```
k, v := cursor.First()
for k != nil && v != nil {
    if string(k) == "key" {
        cursor.Delete()
    }

    k, v = cursor.Next()
}
```
In this above code we'd skip a key/value since the `Delete()` does not return the new `currKey`.

Reply:
But this pseudo code is not what I implemented?

Reply:
Discussed offline, no change needed